pkg/utils/secrets/vpntlsauth/vpn_tlsauth.go

```diff
@@ -0,0 +1,66 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package vpntlsauth
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+
+	secretsutils "github.com/gardener/gardener/pkg/utils/secrets"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+)
+
+const (
+	// SecretNameTLSAuth is the name of seed server tlsauth Secret.
+	SecretNameTLSAuth = "vpn-seed-server-tlsauth" // #nosec G101 -- No credential.
+)
+
+// VPNTLSAuthConfigFromSecret is a configuration for a VPN TLS auth secret with the tlsauth key itself as part
+// of the configuration.
+type VPNTLSAuthConfigFromSecret struct {
+	Name string
+	Data map[string][]byte
+}
+
+var _ secretsutils.ConfigInterface = &VPNTLSAuthConfigFromSecret{}
+var _ secretsutils.DataInterface = &VPNTLSAuthConfigFromSecret{}
+
+// GetName returns the name of the secret.
+func (s *VPNTLSAuthConfigFromSecret) GetName() string {
+	return s.Name
+}
+
+// Generate implements ConfigInterface.
+func (s *VPNTLSAuthConfigFromSecret) Generate() (secretsutils.DataInterface, error) {
+	return s, nil
+}
+
+// SecretData computes the data map which can be used in a Kubernetes secret.
+func (s *VPNTLSAuthConfigFromSecret) SecretData() map[string][]byte {
+	return s.Data
+}
+
+// GenerateSecret generates a VPN TLS auth secret using the provided secrets manager.
+// It is used for two-staged generation of tlsauth secret to include the tlsauth key in the secret name hash.
+func GenerateSecret(ctx context.Context, secretsManager secretsmanager.Interface) (*corev1.Secret, error) {
+	// generate a secret with the tlsauth key
+	secretTLSAuthIntermediate, err := secretsManager.Generate(ctx, &secretsutils.VPNTLSAuthConfig{
+		Name: SecretNameTLSAuth,
+	}, secretsmanager.Rotate(secretsmanager.InPlace))
+	if err != nil {
+		return nil, err
+	}
+
+	// use the secret to get a secret with same data but including the tlsauth key itself in name hash
+	secretTLSAuth, err := secretsManager.Generate(ctx, &VPNTLSAuthConfigFromSecret{
+		Name: secretTLSAuthIntermediate.Name,
+		Data: secretTLSAuthIntermediate.Data,
+	}, secretsmanager.Rotate(secretsmanager.InPlace))
+	if err != nil {
+		return nil, err
+	}
+	return secretTLSAuth, nil
```

Comment:
The first call of `secretsManager.Generate` either gets the key from an existing secret or generates it. Therefore it cannot be in the input secret config.

Reply:
Okay, I now understand why. However, I'm still not happy about generating two secrets 😅 

If you want to use two secrets, you could use `kubernetesutils.MakeUnique` in the vpn-seed-server component like for the shoot case ([ref](https://github.com/gardener/gardener/blob/539913d05582f88b80ea99cc53f2487aebaeeeab/pkg/component/networking/vpn/shoot/shoot.go#L353-L382)).

Given the problem that vpn-seed-server deployment does not notice when the tls-auth key changes, you could also add a label with something like the hashed creation timestamp of `vpn-seed-server-tlsauth-*` secret to its pod template. This would keep the number of secrets lower.

Reply:
+1

However, both approaches sound very hacky to me.

Circling back to the PR description, it says:
> In edge cases, e.g. if the vpn-seed-server-tlsauth-* secret is deleted,

What exactly is this edge-case and wouldn't it be an option to consider this as a field in the config?

Reply:
I'd also prefer to understand more about this edge case first 👍🏻

Reply:
> What exactly is this edge-case and wouldn't it be an option to consider this as a field in the config?

One edge-case is manual deletion of the tlsauth secret in the control plane. An additional field in the config does not help here.
It may not be the edge case we have seen on the productive clusters multiple times, but a reconciliation should be able to repair it. Currently this is not the case.

Reply:
Manual deletion of a human operator? Aren't there many more issues when something like this happens, e.g., what if a CA secret is manually deleted? Will this automatically repair for all components when a reconciliation is triggered?